{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SCCAF clustering for Cai 2020 + 2022 healthy PBMCs\n",
    "\n",
    "**Objective**: Run SCCAF clustering in separate environment to solve dependency issues\n",
    "\n",
    "- **Developed by**: Mairi McClean\n",
    "- **Institute of Computational Biology - Computational Health Centre - Helmholtz Munich**\n",
    "- v230316"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use `SCCAF` to select `leiden` resolution\n",
    "\n",
    "> What is SCCAF? \"Single Cell Clustering Assessment Framework\"\n",
    ">> It is a way to cluster cells based on gene expression; it uses repeat applications of clustering and ML models to generate gene expression profiles - it identifies distinct cell groups and a weighted list of feature genes for each group\n",
    "\n",
    "> What is the architecture/algorithm of the model?\n",
    ">> ML [logistic regression, random forest, Gaussian process classification, support vector machine and decision tree] and 5-fold CV"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mairi.mcclean/mambaforge/envs/scanpy_scvi_local/lib/python3.10/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import anndata\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import seaborn as sns\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy.random as random\n",
    "\n",
    "\n",
    "from umap import UMAP\n",
    "import warnings; warnings.simplefilter('ignore')\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import SCCAF as sccaf\n",
    "from SCCAF import SCCAF_assessment, plot_roc\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in adata objects"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### adata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22049 × 8000\n",
       "    obs: 'study', 'individual', 'sample', 'tissue', 'donor', 'age', 'gender', 'status', 'data_type', 'centre', 'version', 'batch', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'percent_mt2', 'n_counts', 'percent_chrY', 'XIST-counts', 'S_score', 'G2M_score', 'predicted_doublets', 'object', 'protocol', 'dataset', '_scvi_batch', '_scvi_labels'\n",
       "    var: 'gene_id', 'mt', 'ribo', 'n_cells_by_counts-caiy2020', 'mean_counts-caiy2020', 'pct_dropout_by_counts-caiy2020', 'total_counts-caiy2020', 'n_cells_by_counts-caiy2022', 'mean_counts-caiy2022', 'pct_dropout_by_counts-caiy2022', 'total_counts-caiy2022', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'highly_variable_nbatches'\n",
       "    uns: '_scvi_manager_uuid', '_scvi_uuid', 'dataset_colors', 'hvg', 'individual_colors', 'neighbors', 'sample_colors', 'status_colors', 'study_colors', 'tissue_colors', 'umap'\n",
       "    obsm: 'X_scVI', 'X_umap', '_scvi_extra_categorical_covs', '_scvi_extra_continuous_covs'\n",
       "    layers: 'counts', 'sqrt_norm'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad('/Volumes/LaCie/data_lake/Mairi_example/processed_files/scvi/pre_sccaf/Cai_healthy_scRNA_PBMC_mm2303016_pre_sccaf_adata.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>individual</th>\n",
       "      <th>sample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>donor</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>status</th>\n",
       "      <th>data_type</th>\n",
       "      <th>centre</th>\n",
       "      <th>...</th>\n",
       "      <th>percent_chrY</th>\n",
       "      <th>XIST-counts</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>object</th>\n",
       "      <th>protocol</th>\n",
       "      <th>dataset</th>\n",
       "      <th>_scvi_batch</th>\n",
       "      <th>_scvi_labels</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGATGGCGT-H2-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>PBMC_HC_2</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.138408</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.274516</td>\n",
       "      <td>-0.168208</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGATGTGTA-H2-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>PBMC_HC_2</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.162022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.291798</td>\n",
       "      <td>-0.157942</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCAACGGT-H2-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>PBMC_HC_2</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.174764</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.315739</td>\n",
       "      <td>0.098983</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCCATCGC-H2-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>PBMC_HC_2</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015828</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037617</td>\n",
       "      <td>0.046320</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGTGATTA-H2-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>PBMC_HC_2</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048020</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.109589</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001693</td>\n",
       "      <td>-0.202441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTCGCGGTT-H1-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>PBMC_HC_1</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>26</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.110193</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.165353</td>\n",
       "      <td>-0.250572</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTCGCTTCT-H1-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>PBMC_HC_1</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>26</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043309</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.177318</td>\n",
       "      <td>-0.173327</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTCGAATC-H1-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>PBMC_HC_1</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>26</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.235875</td>\n",
       "      <td>-0.013744</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTTGGGCC-H1-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>PBMC_HC_1</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>26</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.091659</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.050351</td>\n",
       "      <td>0.022269</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCCATTAT-H1-caiy2020</th>\n",
       "      <td>CaiY_2021</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>PBMC_HC_1</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>SAMN14048019</td>\n",
       "      <td>26</td>\n",
       "      <td>male</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Shenzhen University</td>\n",
       "      <td>...</td>\n",
       "      <td>0.300300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.080008</td>\n",
       "      <td>0.039641</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caiy2020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22049 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  study    individual     sample tissue  \\\n",
       "barcode                                                                   \n",
       "AAACCTGAGATGGCGT-H2-caiy2020  CaiY_2021  SAMN14048020  PBMC_HC_2   PBMC   \n",
       "AAACCTGAGATGTGTA-H2-caiy2020  CaiY_2021  SAMN14048020  PBMC_HC_2   PBMC   \n",
       "AAACCTGAGCAACGGT-H2-caiy2020  CaiY_2021  SAMN14048020  PBMC_HC_2   PBMC   \n",
       "AAACCTGAGCCATCGC-H2-caiy2020  CaiY_2021  SAMN14048020  PBMC_HC_2   PBMC   \n",
       "AAACCTGAGGTGATTA-H2-caiy2020  CaiY_2021  SAMN14048020  PBMC_HC_2   PBMC   \n",
       "...                                 ...           ...        ...    ...   \n",
       "TTTGTCAGTCGCGGTT-H1-caiy2020  CaiY_2021  SAMN14048019  PBMC_HC_1   PBMC   \n",
       "TTTGTCAGTCGCTTCT-H1-caiy2020  CaiY_2021  SAMN14048019  PBMC_HC_1   PBMC   \n",
       "TTTGTCAGTTCGAATC-H1-caiy2020  CaiY_2021  SAMN14048019  PBMC_HC_1   PBMC   \n",
       "TTTGTCAGTTTGGGCC-H1-caiy2020  CaiY_2021  SAMN14048019  PBMC_HC_1   PBMC   \n",
       "TTTGTCATCCCATTAT-H1-caiy2020  CaiY_2021  SAMN14048019  PBMC_HC_1   PBMC   \n",
       "\n",
       "                                     donor age gender   status data_type  \\\n",
       "barcode                                                                    \n",
       "AAACCTGAGATGGCGT-H2-caiy2020  SAMN14048020  30   male  Healthy  scRNAseq   \n",
       "AAACCTGAGATGTGTA-H2-caiy2020  SAMN14048020  30   male  Healthy  scRNAseq   \n",
       "AAACCTGAGCAACGGT-H2-caiy2020  SAMN14048020  30   male  Healthy  scRNAseq   \n",
       "AAACCTGAGCCATCGC-H2-caiy2020  SAMN14048020  30   male  Healthy  scRNAseq   \n",
       "AAACCTGAGGTGATTA-H2-caiy2020  SAMN14048020  30   male  Healthy  scRNAseq   \n",
       "...                                    ...  ..    ...      ...       ...   \n",
       "TTTGTCAGTCGCGGTT-H1-caiy2020  SAMN14048019  26   male  Healthy  scRNAseq   \n",
       "TTTGTCAGTCGCTTCT-H1-caiy2020  SAMN14048019  26   male  Healthy  scRNAseq   \n",
       "TTTGTCAGTTCGAATC-H1-caiy2020  SAMN14048019  26   male  Healthy  scRNAseq   \n",
       "TTTGTCAGTTTGGGCC-H1-caiy2020  SAMN14048019  26   male  Healthy  scRNAseq   \n",
       "TTTGTCATCCCATTAT-H1-caiy2020  SAMN14048019  26   male  Healthy  scRNAseq   \n",
       "\n",
       "                                           centre  ... percent_chrY  \\\n",
       "barcode                                            ...                \n",
       "AAACCTGAGATGGCGT-H2-caiy2020  Shenzhen University  ...     0.138408   \n",
       "AAACCTGAGATGTGTA-H2-caiy2020  Shenzhen University  ...     0.162022   \n",
       "AAACCTGAGCAACGGT-H2-caiy2020  Shenzhen University  ...     0.174764   \n",
       "AAACCTGAGCCATCGC-H2-caiy2020  Shenzhen University  ...     0.015828   \n",
       "AAACCTGAGGTGATTA-H2-caiy2020  Shenzhen University  ...     0.109589   \n",
       "...                                           ...  ...          ...   \n",
       "TTTGTCAGTCGCGGTT-H1-caiy2020  Shenzhen University  ...     0.110193   \n",
       "TTTGTCAGTCGCTTCT-H1-caiy2020  Shenzhen University  ...     0.043309   \n",
       "TTTGTCAGTTCGAATC-H1-caiy2020  Shenzhen University  ...     0.040551   \n",
       "TTTGTCAGTTTGGGCC-H1-caiy2020  Shenzhen University  ...     0.091659   \n",
       "TTTGTCATCCCATTAT-H1-caiy2020  Shenzhen University  ...     0.300300   \n",
       "\n",
       "                             XIST-counts   S_score  G2M_score  \\\n",
       "barcode                                                         \n",
       "AAACCTGAGATGGCGT-H2-caiy2020         0.0 -0.274516  -0.168208   \n",
       "AAACCTGAGATGTGTA-H2-caiy2020         0.0 -0.291798  -0.157942   \n",
       "AAACCTGAGCAACGGT-H2-caiy2020         0.0 -0.315739   0.098983   \n",
       "AAACCTGAGCCATCGC-H2-caiy2020         0.0  0.037617   0.046320   \n",
       "AAACCTGAGGTGATTA-H2-caiy2020         0.0  0.001693  -0.202441   \n",
       "...                                  ...       ...        ...   \n",
       "TTTGTCAGTCGCGGTT-H1-caiy2020         0.0 -0.165353  -0.250572   \n",
       "TTTGTCAGTCGCTTCT-H1-caiy2020         0.0 -0.177318  -0.173327   \n",
       "TTTGTCAGTTCGAATC-H1-caiy2020         0.0 -0.235875  -0.013744   \n",
       "TTTGTCAGTTTGGGCC-H1-caiy2020         0.0 -0.050351   0.022269   \n",
       "TTTGTCATCCCATTAT-H1-caiy2020         0.0  0.080008   0.039641   \n",
       "\n",
       "                              predicted_doublets  object  protocol   dataset  \\\n",
       "barcode                                                                        \n",
       "AAACCTGAGATGGCGT-H2-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "AAACCTGAGATGTGTA-H2-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "AAACCTGAGCAACGGT-H2-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "AAACCTGAGCCATCGC-H2-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "AAACCTGAGGTGATTA-H2-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "...                                          ...     ...       ...       ...   \n",
       "TTTGTCAGTCGCGGTT-H1-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "TTTGTCAGTCGCTTCT-H1-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "TTTGTCAGTTCGAATC-H1-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "TTTGTCAGTTTGGGCC-H1-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "TTTGTCATCCCATTAT-H1-caiy2020                 0.0     NaN       NaN  caiy2020   \n",
       "\n",
       "                              _scvi_batch  _scvi_labels  \n",
       "barcode                                                  \n",
       "AAACCTGAGATGGCGT-H2-caiy2020            0             0  \n",
       "AAACCTGAGATGTGTA-H2-caiy2020            0             0  \n",
       "AAACCTGAGCAACGGT-H2-caiy2020            0             0  \n",
       "AAACCTGAGCCATCGC-H2-caiy2020            0             0  \n",
       "AAACCTGAGGTGATTA-H2-caiy2020            0             0  \n",
       "...                                   ...           ...  \n",
       "TTTGTCAGTCGCGGTT-H1-caiy2020            0             0  \n",
       "TTTGTCAGTCGCTTCT-H1-caiy2020            0             0  \n",
       "TTTGTCAGTTCGAATC-H1-caiy2020            0             0  \n",
       "TTTGTCAGTTTGGGCC-H1-caiy2020            0             0  \n",
       "TTTGTCATCCCATTAT-H1-caiy2020            0             0  \n",
       "\n",
       "[22049 rows x 31 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### caiy_healthy object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22049 × 61533\n",
       "    obs: 'study', 'individual', 'sample', 'tissue', 'donor', 'age', 'gender', 'status', 'data_type', 'centre', 'version', 'batch', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'percent_mt2', 'n_counts', 'percent_chrY', 'XIST-counts', 'S_score', 'G2M_score', 'predicted_doublets', 'object', 'protocol', 'dataset'\n",
       "    var: 'gene_id', 'mt', 'ribo', 'n_cells_by_counts-caiy2020', 'mean_counts-caiy2020', 'pct_dropout_by_counts-caiy2020', 'total_counts-caiy2020', 'n_cells_by_counts-caiy2022', 'mean_counts-caiy2022', 'pct_dropout_by_counts-caiy2022', 'total_counts-caiy2022'\n",
       "    layers: 'counts', 'sqrt_norm'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "caiy_healthy = sc.read_h5ad('/Volumes/LaCie/data_lake/Mairi_example/processed_files/scvi/pre_sccaf/Cai_healthy_scRNA_PBMC_mm2303016_pre_sccaf_caiy_healthy.h5ad')\n",
    "caiy_healthy"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'louvain'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/mambaforge/envs/scanpy_scvi_local/lib/python3.10/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3801\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[1;32m   3803\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/mambaforge/envs/scanpy_scvi_local/lib/python3.10/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/mambaforge/envs/scanpy_scvi_local/lib/python3.10/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'louvain'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m y_prob, y_pred, y_test, clf, cvsm, acc \u001b[39m=\u001b[39m SCCAF_assessment(adata\u001b[39m.\u001b[39mX, adata\u001b[39m.\u001b[39;49mobs[\u001b[39m'\u001b[39;49m\u001b[39mlouvain\u001b[39;49m\u001b[39m'\u001b[39;49m], n \u001b[39m=\u001b[39m \u001b[39m100\u001b[39m)\n",
      "File \u001b[0;32m~/mambaforge/envs/scanpy_scvi_local/lib/python3.10/site-packages/pandas/core/frame.py:3807\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3805\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mnlevels \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[1;32m   3806\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3807\u001b[0m indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mget_loc(key)\n\u001b[1;32m   3808\u001b[0m \u001b[39mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3809\u001b[0m     indexer \u001b[39m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/mambaforge/envs/scanpy_scvi_local/lib/python3.10/site-packages/pandas/core/indexes/base.py:3804\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3802\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine\u001b[39m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3803\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m-> 3804\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[1;32m   3806\u001b[0m     \u001b[39m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[39m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3808\u001b[0m     \u001b[39m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3809\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'louvain'"
     ]
    }
   ],
   "source": [
    "y_prob, y_pred, y_test, clf, cvsm, acc = SCCAF_assessment(adata.X, adata.obs['louvain'], n = 100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution = 0.7, random_state = 1786)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_roc(y_prob, y_test, clf, cvsm = cvsm, acc = acc)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, frameon = False, color = ['leiden', 'status', 'CD74'], size = 0.8, legend_fontsize = 5, legend_loc = 'on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, frameon = False, color = ['leiden', 'status', 'tissue', 'ADH7', 'CDH1', 'CD74', 'CD3E', 'MUC20', 'DUSP4', 'FOXJ1', 'MUC1', 'FOXI1'], size = 1, legend_fontsize = 5)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export clustered object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata\n",
    "caiy_healthy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making a hybrid anndata object using sections from both original anndata object and the cai_tb_gex object\n",
    "adata_export = anndata.AnnData(X = caiy_healthy.X, var = caiy_healthy.var, obs = adata.obs, uns = adata.uns, obsm = adata.obsm, layers = caiy_healthy.layers, obsp = adata.obsp)\n",
    "adata_export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_export.write('/Volumes/Lacie/data_lake/Mairi_example/processed_files/scvi/post_sccaf/CaiY_healthy_scRNA_PBMC_mm230316_scVI-clustered.raw.h5ad')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sccaf_local",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
