{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## scANVI analysis for healthy PBMC pilot study (Cai 2020 and Cai 2022)\n",
    "\n",
    "**Objective**: Run scANVI analysis for label transfer for healthy PBMCs [Cai 2020 and Cai 2022]\n",
    "\n",
    "\n",
    "- **Developed by**: Mairi McClean\n",
    "\n",
    "- **Institute of Computational Biology - Computational Health Centre - Helmholtz Munich**\n",
    "\n",
    "- v230317\n",
    "\n",
    "- Following this tutorial: https://docs.scvi-tools.org/en/stable/tutorials/notebooks/scarches_scvi_tools.html\n",
    "> \"This particular workflow is useful in the case where a model is trained on some data (called reference here) and new samples are received (called query).\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "text\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sanity check\n",
    "import os \n",
    "os.write(1, b\"text\\n\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import modules "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mairi.mcclean/mambaforge/envs/scanvi_local/lib/python3.10/site-packages/scvi_colab/_core.py:41: UserWarning: \n",
      "                Not currently in Google Colab environment.\n",
      "\n",
      "                Please run with `run_outside_colab=True` to override.\n",
      "\n",
      "                Returning with no further action.\n",
      "                \n",
      "  warn(\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "from scvi_colab import install\n",
    "\n",
    "install()\n",
    "\n",
    "IN_COLAB = \"google.colab\" in sys.modules\n",
    "if IN_COLAB:\n",
    "    %pip install --quiet scrublet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n",
      "/Users/mairi.mcclean/mambaforge/envs/scanvi_local/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/Users/mairi.mcclean/mambaforge/envs/scanvi_local/lib/python3.10/site-packages/flax/struct.py:132: FutureWarning: jax.tree_util.register_keypaths is deprecated, and will be removed in a future release. Please use `register_pytree_with_keys()` instead.\n",
      "  jax.tree_util.register_keypaths(data_clz, keypaths)\n",
      "/Users/mairi.mcclean/mambaforge/envs/scanvi_local/lib/python3.10/site-packages/flax/struct.py:132: FutureWarning: jax.tree_util.register_keypaths is deprecated, and will be removed in a future release. Please use `register_pytree_with_keys()` instead.\n",
      "  jax.tree_util.register_keypaths(data_clz, keypaths)\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import warnings\n",
    "\n",
    "import anndata\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scrublet as scr\n",
    "import scvi\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 94705\n"
     ]
    }
   ],
   "source": [
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "\n",
    "\n",
    "sc.set_figure_params(figsize=(4, 4))\n",
    "scvi.settings.seed = 94705\n",
    "\n",
    "%config InlineBackend.print_figure_kwargs={'facecolor' : \"w\"}\n",
    "%config InlineBackend.figure_format='retina'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in and prepare data\n",
    "- Following steps from scVI portion of tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 422220 × 33751\n",
       "    obs: 'patient_id', 'Ethnicity', 'BMI', 'annotation_broad', 'annotation_detailed', 'annotation_detailed_fullNames', 'Age_group', 'COVID_severity', 'COVID_status', 'Group', 'Sex', 'Smoker', 'sample_id', 'sequencing_library', 'Protein_modality_weight'\n",
       "    var: 'name'\n",
       "    obsm: 'X_ umap (wnn derived)', 'X_umap (after harmony ADT)', 'X_umap (after harmony RNA)', 'X_umap (before harmony ADT)', 'X_umap (before harmony RNA)'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ref = sc.read_h5ad('/Volumes/Lacie/data_lake/Mairi_example/INBOX/sc_downloads/yoshida_2021/meyer_nikolic_covid_pbmc.cellxgene.20210813.h5ad')\n",
    "adata_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>BMI</th>\n",
       "      <th>annotation_broad</th>\n",
       "      <th>annotation_detailed</th>\n",
       "      <th>annotation_detailed_fullNames</th>\n",
       "      <th>Age_group</th>\n",
       "      <th>COVID_severity</th>\n",
       "      <th>COVID_status</th>\n",
       "      <th>Group</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sequencing_library</th>\n",
       "      <th>Protein_modality_weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAGGTT-1</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte CD14</td>\n",
       "      <td>Classical monocyte</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>AN5</td>\n",
       "      <td>CV001_KM10202384-CV001_KM10202394</td>\n",
       "      <td>0.359517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTCCCG-1</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>AN5</td>\n",
       "      <td>CV001_KM10202384-CV001_KM10202394</td>\n",
       "      <td>0.577522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATGGGT-1</th>\n",
       "      <td>AN3</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>AN3</td>\n",
       "      <td>CV001_KM10202384-CV001_KM10202394</td>\n",
       "      <td>0.369143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAGTAG-1</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD8+</td>\n",
       "      <td>T CD8 naive</td>\n",
       "      <td>T CD8 naive</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>AN5</td>\n",
       "      <td>CV001_KM10202384-CV001_KM10202394</td>\n",
       "      <td>0.785563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGCGTC-1</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 naive</td>\n",
       "      <td>T CD4 naive</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Female</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>AN5</td>\n",
       "      <td>CV001_KM10202384-CV001_KM10202394</td>\n",
       "      <td>0.564174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S28_TTTGTCAGTTCTGTTT-1</th>\n",
       "      <td>PC9</td>\n",
       "      <td>EUR</td>\n",
       "      <td>27.17</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Post-COVID-19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>PC9</td>\n",
       "      <td>CV001_KM9294396-CV001_KM9294404</td>\n",
       "      <td>0.429398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S28_TTTGTCATCAACCAAC-1</th>\n",
       "      <td>PC9</td>\n",
       "      <td>EUR</td>\n",
       "      <td>27.17</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte CD14</td>\n",
       "      <td>Classical monocyte</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Post-COVID-19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>PC9</td>\n",
       "      <td>CV001_KM9294396-CV001_KM9294404</td>\n",
       "      <td>0.677910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S28_TTTGTCATCATTATCC-1</th>\n",
       "      <td>PC9</td>\n",
       "      <td>EUR</td>\n",
       "      <td>27.17</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte CD14</td>\n",
       "      <td>Classical monocyte</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Post-COVID-19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>PC9</td>\n",
       "      <td>CV001_KM9294396-CV001_KM9294404</td>\n",
       "      <td>0.422796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S28_TTTGTCATCCTATGTT-1</th>\n",
       "      <td>PC9</td>\n",
       "      <td>EUR</td>\n",
       "      <td>27.17</td>\n",
       "      <td>DC</td>\n",
       "      <td>pDC</td>\n",
       "      <td>pDC</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Post-COVID-19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>PC9</td>\n",
       "      <td>CV001_KM9294396-CV001_KM9294404</td>\n",
       "      <td>0.471905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S28_TTTGTCATCTCAAGTG-1</th>\n",
       "      <td>PC2</td>\n",
       "      <td>EUR</td>\n",
       "      <td>36.9</td>\n",
       "      <td>T CD8+</td>\n",
       "      <td>T CD8 CTL</td>\n",
       "      <td>T CD8 CTL</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Post-COVID-19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Male</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>PC2</td>\n",
       "      <td>CV001_KM9294396-CV001_KM9294404</td>\n",
       "      <td>0.313640</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422220 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   patient_id Ethnicity  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        AN3       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...        AN5       EUR   \n",
       "...                                                       ...       ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                                    PC9       EUR   \n",
       "S28_TTTGTCATCAACCAAC-1                                    PC9       EUR   \n",
       "S28_TTTGTCATCATTATCC-1                                    PC9       EUR   \n",
       "S28_TTTGTCATCCTATGTT-1                                    PC9       EUR   \n",
       "S28_TTTGTCATCTCAAGTG-1                                    PC2       EUR   \n",
       "\n",
       "                                                        BMI annotation_broad  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...  Unknown         Monocyte   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...  Unknown           T CD4+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...  Unknown           T CD4+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...  Unknown           T CD8+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...  Unknown           T CD4+   \n",
       "...                                                     ...              ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                                27.17               NK   \n",
       "S28_TTTGTCATCAACCAAC-1                                27.17         Monocyte   \n",
       "S28_TTTGTCATCATTATCC-1                                27.17         Monocyte   \n",
       "S28_TTTGTCATCCTATGTT-1                                27.17               DC   \n",
       "S28_TTTGTCATCTCAAGTG-1                                 36.9           T CD8+   \n",
       "\n",
       "                                                   annotation_detailed  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...       Monocyte CD14   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...         T CD8 naive   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...         T CD4 naive   \n",
       "...                                                                ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                                              NK   \n",
       "S28_TTTGTCATCAACCAAC-1                                   Monocyte CD14   \n",
       "S28_TTTGTCATCATTATCC-1                                   Monocyte CD14   \n",
       "S28_TTTGTCATCCTATGTT-1                                             pDC   \n",
       "S28_TTTGTCATCTCAAGTG-1                                       T CD8 CTL   \n",
       "\n",
       "                                                   annotation_detailed_fullNames  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...            Classical monocyte   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...                  T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...                  T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...                   T CD8 naive   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...                   T CD4 naive   \n",
       "...                                                                          ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                                                        NK   \n",
       "S28_TTTGTCATCAACCAAC-1                                        Classical monocyte   \n",
       "S28_TTTGTCATCATTATCC-1                                        Classical monocyte   \n",
       "S28_TTTGTCATCCTATGTT-1                                                       pDC   \n",
       "S28_TTTGTCATCTCAAGTG-1                                                 T CD8 CTL   \n",
       "\n",
       "                                                   Age_group COVID_severity  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...     Adult        Healthy   \n",
       "...                                                      ...            ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                                 Adult         Severe   \n",
       "S28_TTTGTCATCAACCAAC-1                                 Adult         Severe   \n",
       "S28_TTTGTCATCATTATCC-1                                 Adult         Severe   \n",
       "S28_TTTGTCATCCTATGTT-1                                 Adult         Severe   \n",
       "S28_TTTGTCATCTCAAGTG-1                                 Adult         Severe   \n",
       "\n",
       "                                                     COVID_status  Group  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...        Healthy  Adult   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        Healthy  Adult   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        Healthy  Adult   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...        Healthy  Adult   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...        Healthy  Adult   \n",
       "...                                                           ...    ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                              Post-COVID-19  Adult   \n",
       "S28_TTTGTCATCAACCAAC-1                              Post-COVID-19  Adult   \n",
       "S28_TTTGTCATCATTATCC-1                              Post-COVID-19  Adult   \n",
       "S28_TTTGTCATCCTATGTT-1                              Post-COVID-19  Adult   \n",
       "S28_TTTGTCATCTCAAGTG-1                              Post-COVID-19  Adult   \n",
       "\n",
       "                                                       Sex      Smoker  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...  Female  Non-smoker   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...  Female  Non-smoker   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...    Male  Non-smoker   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...  Female  Non-smoker   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...  Female  Non-smoker   \n",
       "...                                                    ...         ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                                Male  Non-smoker   \n",
       "S28_TTTGTCATCAACCAAC-1                                Male  Non-smoker   \n",
       "S28_TTTGTCATCATTATCC-1                                Male  Non-smoker   \n",
       "S28_TTTGTCATCCTATGTT-1                                Male  Non-smoker   \n",
       "S28_TTTGTCATCTCAAGTG-1                                Male  Non-smoker   \n",
       "\n",
       "                                                   sample_id  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...       AN5   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...       AN5   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...       AN3   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...       AN5   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...       AN5   \n",
       "...                                                      ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                                   PC9   \n",
       "S28_TTTGTCATCAACCAAC-1                                   PC9   \n",
       "S28_TTTGTCATCATTATCC-1                                   PC9   \n",
       "S28_TTTGTCATCCTATGTT-1                                   PC9   \n",
       "S28_TTTGTCATCTCAAGTG-1                                   PC2   \n",
       "\n",
       "                                                                   sequencing_library  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...  CV001_KM10202384-CV001_KM10202394   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...  CV001_KM10202384-CV001_KM10202394   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...  CV001_KM10202384-CV001_KM10202394   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...  CV001_KM10202384-CV001_KM10202394   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...  CV001_KM10202384-CV001_KM10202394   \n",
       "...                                                                               ...   \n",
       "S28_TTTGTCAGTTCTGTTT-1                                CV001_KM9294396-CV001_KM9294404   \n",
       "S28_TTTGTCATCAACCAAC-1                                CV001_KM9294396-CV001_KM9294404   \n",
       "S28_TTTGTCATCATTATCC-1                                CV001_KM9294396-CV001_KM9294404   \n",
       "S28_TTTGTCATCCTATGTT-1                                CV001_KM9294396-CV001_KM9294404   \n",
       "S28_TTTGTCATCTCAAGTG-1                                CV001_KM9294396-CV001_KM9294404   \n",
       "\n",
       "                                                    Protein_modality_weight  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...                 0.359517  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...                 0.577522  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...                 0.369143  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...                 0.785563  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...                 0.564174  \n",
       "...                                                                     ...  \n",
       "S28_TTTGTCAGTTCTGTTT-1                                             0.429398  \n",
       "S28_TTTGTCATCAACCAAC-1                                             0.677910  \n",
       "S28_TTTGTCATCATTATCC-1                                             0.422796  \n",
       "S28_TTTGTCATCCTATGTT-1                                             0.471905  \n",
       "S28_TTTGTCATCTCAAGTG-1                                             0.313640  \n",
       "\n",
       "[422220 rows x 15 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ref.obs"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Filter query dataset to online include healthy cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy          173684\n",
       "COVID-19         151312\n",
       "Post-COVID-19     97224\n",
       "Name: COVID_status, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ref.obs['COVID_status'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 173684 × 33751\n",
       "    obs: 'patient_id', 'Ethnicity', 'BMI', 'annotation_broad', 'annotation_detailed', 'annotation_detailed_fullNames', 'Age_group', 'COVID_severity', 'COVID_status', 'Group', 'Sex', 'Smoker', 'sample_id', 'sequencing_library', 'Protein_modality_weight'\n",
       "    var: 'name'\n",
       "    obsm: 'X_ umap (wnn derived)', 'X_umap (after harmony ADT)', 'X_umap (after harmony RNA)', 'X_umap (before harmony ADT)', 'X_umap (before harmony RNA)'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ref_new = adata_ref[~adata_ref.obs['COVID_status'].isin(['COVID-19', 'Post-COVID-19']),:]\n",
    "adata_ref_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy    173684\n",
       "Name: COVID_status, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ref_new.obs['COVID_status'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Read in query data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 73146 × 61533\n",
       "    obs: 'study', 'individual', 'sample', 'tissue', 'donor', 'age', 'gender', 'status', 'data_type', 'centre', 'version', 'batch', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'percent_mt2', 'n_counts', 'percent_chrY', 'XIST-counts', 'S_score', 'G2M_score', 'predicted_doublets'\n",
       "    var: 'gene_id', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'\n",
       "    uns: 'donor_colors'\n",
       "    layers: 'counts', 'sqrt_norm'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Note: Query data is that of only Cai, not the Cai+Yoshida data, but clustered following scvi. This is not the correct data object?\n",
    "\n",
    "# adata_query = sc.read_h5ad('/Volumes/Lacie/data_lake/Mairi_example/processed_files/scvi/post_sccaf/CaiY_healthy_scRNA_PBMC_mm230316_scVI-clustered.raw.h5ad')\n",
    "\n",
    "adata_query_2020 = sc.read_h5ad('/Volumes/LaCie/data_lake/Mairi_example/processed_files/abridged_qc/human/Cai2020_scRNA_PBMC_mm230315_qcd.h5ad')\n",
    "adata_query_2020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 25417 × 61533\n",
       "    obs: 'study', 'individual', 'sample', 'tissue', 'donor', 'data_type', 'centre', 'version', 'object', 'protocol', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'percent_mt2', 'n_counts', 'percent_chrY', 'XIST-counts', 'S_score', 'G2M_score', 'predicted_doublets'\n",
       "    var: 'gene_id', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'\n",
       "    uns: 'donor_colors', 'sample_colors'\n",
       "    layers: 'counts', 'sqrt_norm'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_query_2022 = sc.read_h5ad('/Volumes/LaCie/data_lake/Mairi_example/processed_files/abridged_qc/human/Cai2022_scRNA_PBMC_mm230315_qcd.h5ad')\n",
    "adata_query_2022"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Concatenate both query datasets\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_query = adata_query_2020.concatenate(adata_query_2022, join='inner', batch_categories=['2020', '2022'], batch_key='year')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Concatenate anndata objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mairi.mcclean/mambaforge/envs/scanvi_local/lib/python3.10/site-packages/anndata/_core/merge.py:942: UserWarning: Only some AnnData objects have `.raw` attribute, not concatenating `.raw` attributes.\n",
      "  warn(\n"
     ]
    }
   ],
   "source": [
    "adata_healthy = adata_ref_new.concatenate(adata_query, join=\"inner\", batch_categories=['reference', 'query'], batch_key='dataset')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 272247 × 22792\n",
       "    obs: 'patient_id', 'Ethnicity', 'BMI', 'annotation_broad', 'annotation_detailed', 'annotation_detailed_fullNames', 'Age_group', 'COVID_severity', 'COVID_status', 'Group', 'Sex', 'Smoker', 'sample_id', 'sequencing_library', 'Protein_modality_weight', 'study', 'individual', 'sample', 'tissue', 'donor', 'age', 'gender', 'status', 'data_type', 'centre', 'version', 'batch', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'percent_mt2', 'n_counts', 'percent_chrY', 'XIST-counts', 'S_score', 'G2M_score', 'predicted_doublets', 'object', 'protocol', 'year', 'dataset'\n",
       "    var: 'gene_id-query', 'mt-query', 'ribo-query', 'n_cells_by_counts-2020-query', 'mean_counts-2020-query', 'pct_dropout_by_counts-2020-query', 'total_counts-2020-query', 'n_cells_by_counts-2022-query', 'mean_counts-2022-query', 'pct_dropout_by_counts-2022-query', 'total_counts-2022-query', 'name-reference'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_healthy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>BMI</th>\n",
       "      <th>annotation_broad</th>\n",
       "      <th>annotation_detailed</th>\n",
       "      <th>annotation_detailed_fullNames</th>\n",
       "      <th>Age_group</th>\n",
       "      <th>COVID_severity</th>\n",
       "      <th>COVID_status</th>\n",
       "      <th>Group</th>\n",
       "      <th>...</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>percent_chrY</th>\n",
       "      <th>XIST-counts</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>object</th>\n",
       "      <th>protocol</th>\n",
       "      <th>year</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAGGTT-1-reference</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte CD14</td>\n",
       "      <td>Classical monocyte</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTCCCG-1-reference</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATGGGT-1-reference</th>\n",
       "      <td>AN3</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAGTAG-1-reference</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD8+</td>\n",
       "      <td>T CD8 naive</td>\n",
       "      <td>T CD8 naive</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGCGTC-1-reference</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 naive</td>\n",
       "      <td>T CD4 naive</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>746.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.137711</td>\n",
       "      <td>-0.152166</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.130521</td>\n",
       "      <td>0.092987</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.043432</td>\n",
       "      <td>-0.034049</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1340.0</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.236621</td>\n",
       "      <td>0.071757</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>365.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156833</td>\n",
       "      <td>-0.050452</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>272247 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   patient_id Ethnicity  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        AN3       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...        AN5       EUR   \n",
       "...                                                       ...       ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "\n",
       "                                                        BMI annotation_broad  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...  Unknown         Monocyte   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...  Unknown           T CD4+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...  Unknown           T CD4+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...  Unknown           T CD8+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...  Unknown           T CD4+   \n",
       "...                                                     ...              ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...      NaN              NaN   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...      NaN              NaN   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...      NaN              NaN   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...      NaN              NaN   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...      NaN              NaN   \n",
       "\n",
       "                                                   annotation_detailed  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...       Monocyte CD14   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...         T CD8 naive   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...         T CD4 naive   \n",
       "...                                                                ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...                 NaN   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...                 NaN   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...                 NaN   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...                 NaN   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...                 NaN   \n",
       "\n",
       "                                                   annotation_detailed_fullNames  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...            Classical monocyte   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...                  T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...                  T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...                   T CD8 naive   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...                   T CD4 naive   \n",
       "...                                                                          ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...                           NaN   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...                           NaN   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...                           NaN   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...                           NaN   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...                           NaN   \n",
       "\n",
       "                                                   Age_group COVID_severity  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...     Adult        Healthy   \n",
       "...                                                      ...            ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "\n",
       "                                                   COVID_status  Group  ...  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...      Healthy  Adult  ...   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...      Healthy  Adult  ...   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...      Healthy  Adult  ...   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...      Healthy  Adult  ...   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...      Healthy  Adult  ...   \n",
       "...                                                         ...    ...  ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "\n",
       "                                                   n_counts percent_chrY  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...      NaN          NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...      NaN          NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...      NaN          NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...      NaN          NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...      NaN          NaN   \n",
       "...                                                     ...          ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...    746.0     0.000000   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...    369.0     0.000000   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...    348.0     0.000000   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...   1340.0     0.074627   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...    365.0     0.000000   \n",
       "\n",
       "                                                   XIST-counts   S_score  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...         NaN       NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...         NaN       NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...         NaN       NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...         NaN       NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...         NaN       NaN   \n",
       "...                                                        ...       ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...         0.0 -0.137711   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...         0.0  0.130521   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...         0.0 -0.043432   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...         0.0  0.236621   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...         0.0  0.156833   \n",
       "\n",
       "                                                    G2M_score  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...        NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...        NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...        NaN   \n",
       "...                                                       ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...  -0.152166   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...   0.092987   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...  -0.034049   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...   0.071757   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...  -0.050452   \n",
       "\n",
       "                                                   predicted_doublets  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...                NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...                NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...                NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...                NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...                NaN   \n",
       "...                                                               ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "\n",
       "                                                       object       protocol  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...        NaN            NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        NaN            NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        NaN            NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...        NaN            NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...        NaN            NaN   \n",
       "...                                                       ...            ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "\n",
       "                                                    year    dataset  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...   NaN  reference  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...   NaN  reference  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...   NaN  reference  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...   NaN  reference  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...   NaN  reference  \n",
       "...                                                  ...        ...  \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "\n",
       "[272247 rows x 45 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_healthy.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patient_id                       category\n",
      "Ethnicity                        category\n",
      "BMI                              category\n",
      "annotation_broad                 category\n",
      "annotation_detailed              category\n",
      "annotation_detailed_fullNames    category\n",
      "Age_group                        category\n",
      "COVID_severity                   category\n",
      "COVID_status                     category\n",
      "Group                            category\n",
      "Sex                              category\n",
      "Smoker                           category\n",
      "sample_id                        category\n",
      "sequencing_library               category\n",
      "Protein_modality_weight           float64\n",
      "study                              object\n",
      "individual                         object\n",
      "sample                             object\n",
      "tissue                           category\n",
      "donor                              object\n",
      "age                              category\n",
      "gender                           category\n",
      "status                           category\n",
      "data_type                        category\n",
      "centre                           category\n",
      "version                          category\n",
      "batch                            category\n",
      "n_genes                           float64\n",
      "n_genes_by_counts                 float64\n",
      "total_counts                      float32\n",
      "total_counts_mt                   float32\n",
      "pct_counts_mt                     float32\n",
      "total_counts_ribo                 float32\n",
      "pct_counts_ribo                   float32\n",
      "percent_mt2                       float32\n",
      "n_counts                          float32\n",
      "percent_chrY                      float32\n",
      "XIST-counts                       float32\n",
      "S_score                           float64\n",
      "G2M_score                         float64\n",
      "predicted_doublets                float64\n",
      "object                           category\n",
      "protocol                         category\n",
      "year                             category\n",
      "dataset                          category\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(adata_healthy.obs.dtypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T CD4+       49415\n",
       "T CD8+       34110\n",
       "Monocyte     28637\n",
       "B            26132\n",
       "NK           21871\n",
       "T reg         3251\n",
       "T g/d         3183\n",
       "MAIT          2213\n",
       "DC            2151\n",
       "Cycling       1012\n",
       "Platelets      626\n",
       "HPC            416\n",
       "Plasma         352\n",
       "ILC            199\n",
       "RBC            106\n",
       "Baso/Eos        10\n",
       "Name: annotation_broad, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_healthy.obs['annotation_broad'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_healthy.obs[\"annotation_broad\"] = adata_healthy.obs[\"annotation_broad\"].cat.add_categories(\"unknown\").fillna(\"unknown\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_healthy.obs[\"annotation_detailed\"] = adata_healthy.obs[\"annotation_detailed\"].cat.add_categories(\"unknown\").fillna(\"unknown\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_healthy.obs[\"annotation_detailed_fullNames\"] = adata_healthy.obs[\"annotation_detailed_fullNames\"].cat.add_categories(\"unknown\").fillna(\"unknown\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>BMI</th>\n",
       "      <th>annotation_broad</th>\n",
       "      <th>annotation_detailed</th>\n",
       "      <th>annotation_detailed_fullNames</th>\n",
       "      <th>Age_group</th>\n",
       "      <th>COVID_severity</th>\n",
       "      <th>COVID_status</th>\n",
       "      <th>Group</th>\n",
       "      <th>...</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>percent_chrY</th>\n",
       "      <th>XIST-counts</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>predicted_doublets</th>\n",
       "      <th>object</th>\n",
       "      <th>protocol</th>\n",
       "      <th>year</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAGGTT-1-reference</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte CD14</td>\n",
       "      <td>Classical monocyte</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTCCCG-1-reference</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATGGGT-1-reference</th>\n",
       "      <td>AN3</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>T CD4 helper</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAGTAG-1-reference</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD8+</td>\n",
       "      <td>T CD8 naive</td>\n",
       "      <td>T CD8 naive</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGCGTC-1-reference</th>\n",
       "      <td>AN5</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>T CD4+</td>\n",
       "      <td>T CD4 naive</td>\n",
       "      <td>T CD4 naive</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>746.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.137711</td>\n",
       "      <td>-0.152166</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.130521</td>\n",
       "      <td>0.092987</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.043432</td>\n",
       "      <td>-0.034049</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1340.0</td>\n",
       "      <td>0.074627</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.236621</td>\n",
       "      <td>0.071757</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-query</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>365.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156833</td>\n",
       "      <td>-0.050452</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HRS100513</td>\n",
       "      <td>pbmc_scRNAseq</td>\n",
       "      <td>2022</td>\n",
       "      <td>query</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>272247 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   patient_id Ethnicity  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        AN3       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...        AN5       EUR   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...        AN5       EUR   \n",
       "...                                                       ...       ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...        NaN       NaN   \n",
       "\n",
       "                                                        BMI annotation_broad  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...  Unknown         Monocyte   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...  Unknown           T CD4+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...  Unknown           T CD4+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...  Unknown           T CD8+   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...  Unknown           T CD4+   \n",
       "...                                                     ...              ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...      NaN          unknown   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...      NaN          unknown   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...      NaN          unknown   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...      NaN          unknown   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...      NaN          unknown   \n",
       "\n",
       "                                                   annotation_detailed  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...       Monocyte CD14   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...         T CD8 naive   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...         T CD4 naive   \n",
       "...                                                                ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...             unknown   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...             unknown   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...             unknown   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...             unknown   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...             unknown   \n",
       "\n",
       "                                                   annotation_detailed_fullNames  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...            Classical monocyte   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...                  T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...                  T CD4 helper   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...                   T CD8 naive   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...                   T CD4 naive   \n",
       "...                                                                          ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...                       unknown   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...                       unknown   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...                       unknown   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...                       unknown   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...                       unknown   \n",
       "\n",
       "                                                   Age_group COVID_severity  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...     Adult        Healthy   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...     Adult        Healthy   \n",
       "...                                                      ...            ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...       NaN            NaN   \n",
       "\n",
       "                                                   COVID_status  Group  ...  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...      Healthy  Adult  ...   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...      Healthy  Adult  ...   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...      Healthy  Adult  ...   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...      Healthy  Adult  ...   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...      Healthy  Adult  ...   \n",
       "...                                                         ...    ...  ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...          NaN    NaN  ...   \n",
       "\n",
       "                                                   n_counts percent_chrY  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...      NaN          NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...      NaN          NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...      NaN          NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...      NaN          NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...      NaN          NaN   \n",
       "...                                                     ...          ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...    746.0     0.000000   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...    369.0     0.000000   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...    348.0     0.000000   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...   1340.0     0.074627   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...    365.0     0.000000   \n",
       "\n",
       "                                                   XIST-counts   S_score  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...         NaN       NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...         NaN       NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...         NaN       NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...         NaN       NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...         NaN       NaN   \n",
       "...                                                        ...       ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...         0.0 -0.137711   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...         0.0  0.130521   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...         0.0 -0.043432   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...         0.0  0.236621   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...         0.0  0.156833   \n",
       "\n",
       "                                                    G2M_score  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...        NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...        NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...        NaN   \n",
       "...                                                       ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...  -0.152166   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...   0.092987   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...  -0.034049   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...   0.071757   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...  -0.050452   \n",
       "\n",
       "                                                   predicted_doublets  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...                NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...                NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...                NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...                NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...                NaN   \n",
       "...                                                               ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...                0.0   \n",
       "\n",
       "                                                       object       protocol  \\\n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...        NaN            NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...        NaN            NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...        NaN            NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...        NaN            NaN   \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...        NaN            NaN   \n",
       "...                                                       ...            ...   \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...  HRS100513  pbmc_scRNAseq   \n",
       "\n",
       "                                                    year    dataset  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGGCAG...   NaN  reference  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGAGTGTC...   NaN  reference  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGCAGATG...   NaN  reference  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTATAG...   NaN  reference  \n",
       "CV001_KM10202384-CV001_KM10202394_AAACCTGGTGTGC...   NaN  reference  \n",
       "...                                                  ...        ...  \n",
       "TTTGTCAGTCCATCCT-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "TTTGTCAGTGCTGTAT-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "TTTGTCAGTGTGGCTC-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "TTTGTCATCGTTTATC-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "TTTGTCATCTTACCTA-HRS100513-pbmc_scRNAseq-2022-q...  2022      query  \n",
       "\n",
       "[272247 rows x 45 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_healthy.obs"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Rename cell type labels column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_healthy.obs[\"labels_scanvi\"] = adata_healthy.obs[\"annotation_broad\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unknown      98563\n",
       "T CD4+       49415\n",
       "T CD8+       34110\n",
       "Monocyte     28637\n",
       "B            26132\n",
       "NK           21871\n",
       "T reg         3251\n",
       "T g/d         3183\n",
       "MAIT          2213\n",
       "DC            2151\n",
       "Cycling       1012\n",
       "Platelets      626\n",
       "HPC            416\n",
       "Plasma         352\n",
       "ILC            199\n",
       "RBC            106\n",
       "Baso/Eos        10\n",
       "Name: labels_scanvi, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_healthy.obs['labels_scanvi'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Filter genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 272247 × 22792\n",
       "    obs: 'patient_id', 'Ethnicity', 'BMI', 'annotation_broad', 'annotation_detailed', 'annotation_detailed_fullNames', 'Age_group', 'COVID_severity', 'COVID_status', 'Group', 'Sex', 'Smoker', 'sample_id', 'sequencing_library', 'Protein_modality_weight', 'study', 'individual', 'sample', 'tissue', 'donor', 'age', 'gender', 'status', 'data_type', 'centre', 'version', 'batch', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'percent_mt2', 'n_counts', 'percent_chrY', 'XIST-counts', 'S_score', 'G2M_score', 'predicted_doublets', 'object', 'protocol', 'year', 'dataset', 'labels_scanvi'\n",
       "    var: 'gene_id-query', 'mt-query', 'ribo-query', 'n_cells_by_counts-2020-query', 'mean_counts-2020-query', 'pct_dropout_by_counts-2020-query', 'total_counts-2020-query', 'n_cells_by_counts-2022-query', 'mean_counts-2022-query', 'pct_dropout_by_counts-2022-query', 'total_counts-2022-query', 'name-reference'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save counts in layer\n",
    "\n",
    "adata_healthy.layers['counts'] = adata_healthy.X.copy()\n",
    "adata_healthy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mairi.mcclean/mambaforge/envs/scanvi_local/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:62: UserWarning: `flavor='seurat_v3'` expects raw count data, but non-integers were found.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Note: For best practices of how/when to perform feature selection, \n",
    "# please refer to the model-specific tutorial. For scVI, we recommend anywhere from 1,000 to 10,000 HVGs, but it will be context-dependent.\n",
    "\n",
    "sc.pp.highly_variable_genes(\n",
    "    adata_healthy,\n",
    "    n_top_genes=1200,\n",
    "    subset=True,\n",
    "    layer=\"counts\",\n",
    "    flavor=\"seurat_v3\",\n",
    "    batch_key=\"dataset\",\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Set labels and prepare adata subsets for model training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Taken from Anna's notebook\n",
    "\n",
    "condition_key = 'dataset'\n",
    "cell_type_key = 'labels_scanvi'\n",
    "target_conditions = ['query']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:In order to use the mouse gastrulation seqFISH datsets, please install squidpy (see https://github.com/scverse/squidpy).\n",
      "WARNING:root:In order to use sagenet models, please install pytorch geometric (see https://pytorch-geometric.readthedocs.io) and \n",
      " captum (see https://github.com/pytorch/captum).\n",
      "WARNING:root:mvTCR is not installed. To use mvTCR models, please install it first using \"pip install mvtcr\"\n",
      "WARNING:root:multigrate is not installed. To use multigrate models, please install it first using \"pip install multigrate\".\n"
     ]
    }
   ],
   "source": [
    "from scarches.dataset.trvae.data_handling import remove_sparsity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'scarches'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[31], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m# Taken from Anna's notebook\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mscarches\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mdataset\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mtrvae\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mdata_handling\u001b[39;00m \u001b[39mimport\u001b[39;00m remove_sparsity\n\u001b[1;32m      5\u001b[0m adata_healthy \u001b[39m=\u001b[39m remove_sparsity(adata_healthy)\n\u001b[1;32m      6\u001b[0m source_adata \u001b[39m=\u001b[39m adata_healthy[\u001b[39m~\u001b[39madata_healthy\u001b[39m.\u001b[39mobs[condition_key]\u001b[39m.\u001b[39misin(target_conditions)]\u001b[39m.\u001b[39mcopy()\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'scarches'"
     ]
    }
   ],
   "source": [
    "# Taken from Anna's notebook\n",
    "\n",
    "adata_healthy = remove_sparsity(adata_healthy)\n",
    "source_adata = adata_healthy[~adata_healthy.obs[condition_key].isin(target_conditions)].copy()\n",
    "target_adata = adata_healthy[adata_healthy.obs[condition_key].isin(target_conditions)].copy()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train reference\n",
    "\n",
    "> From tutorial: SCANVI tends to perform better in situations where it has been initialized using a pre-trained SCVI model. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- scVI model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scvi.model.SCVI.setup_anndata(adata_ref_new, batch_key=\"sequencing_library\", layer=\"counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Added for sarches params following tutorial; check if we require it?\n",
    "arches_params = dict(\n",
    "    use_layer_norm=\"both\",\n",
    "    use_batch_norm=\"none\",\n",
    "    encode_covariates=True,\n",
    "    dropout_rate=0.2,\n",
    "    n_layers=2,\n",
    ")\n",
    "\n",
    "scvi_ref_new_model = scvi.model.SCVI(adata_ref_new, **arches_params)\n",
    "scvi_ref_new_model.train()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- scANVI model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unlabeled category does not exist in adata.obs[labels_key]\n",
    "# so all cells are treated as labeled\n",
    "vae_ref_scanvi = scvi.model.SCANVI.from_scvi_model(\n",
    "    scvi_ref_new_model,\n",
    "    unlabeled_category=\"Unknown\",\n",
    "    labels_key=\"labels_scanvi\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vae_ref_scanvi.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ref_new.obsm[\"X_scANVI\"] = vae_ref_scanvi.get_latent_representation()\n",
    "sc.pp.neighbors(adata_ref_new, use_rep=\"X_scANVI\")\n",
    "sc.tl.leiden(adata_ref_new)\n",
    "sc.tl.umap(adata_ref_new)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata_ref_new,\n",
    "    color=[\"sequencing_library\", \"annotation_detailed\"],\n",
    "    frameon=False,\n",
    "    ncols=1,\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Update with query"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question: Do you cluster your query data before entering it into the scanvi model?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_path_scan = \"/Volumes/LaCie/data_lake/Mairi_example/processed_files/scanvi/models/\"\n",
    "vae_ref_scanvi.save(dir_path_scan, overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# again a no-op in this tutorial, but good practice to use\n",
    "scvi.model.SCANVI.prepare_query_anndata(adata_query, dir_path_scan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vae_q = scvi.model.SCANVI.load_query_data(\n",
    "    adata_query,\n",
    "    dir_path_scan,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vae_q.train(\n",
    "    max_epochs=100,\n",
    "    plan_kwargs=dict(weight_decay=0.0),\n",
    "    check_val_every_n_epoch=10,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_query.obsm[\"X_scANVI\"] = vae_q.get_latent_representation()\n",
    "adata_query.obs[\"predictions\"] = vae_q.predict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = adata_query.obs.groupby([\"celltype\", \"predictions\"]).size().unstack(fill_value=0)\n",
    "norm_df = df / df.sum(axis=0)\n",
    "\n",
    "plt.figure(figsize=(8, 8))\n",
    "_ = plt.pcolor(norm_df)\n",
    "_ = plt.xticks(np.arange(0.5, len(df.columns), 1), df.columns, rotation=90)\n",
    "_ = plt.yticks(np.arange(0.5, len(df.index), 1), df.index)\n",
    "plt.xlabel(\"Predicted\")\n",
    "plt.ylabel(\"Observed\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analyze reference and query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_full = adata_query.concatenate(adata_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_full.obs.batch.cat.rename_categories([\"Query\", \"Reference\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_predictions = vae_q.predict(adata_full)\n",
    "print(f\"Acc: {np.mean(full_predictions == adata_full.obs.celltype)}\")\n",
    "\n",
    "adata_full.obs[\"predictions\"] = full_predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata_full, use_rep=\"X_scANVI\")\n",
    "sc.tl.leiden(adata_full)\n",
    "sc.tl.umap(adata_full)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata_full,\n",
    "    color=[\"tech\", \"celltype\"],\n",
    "    frameon=False,\n",
    "    ncols=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sc.pl.umap(\n",
    "    adata_full,\n",
    "    frameon=False,\n",
    "    show=False,\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata_full[: adata_query.n_obs],\n",
    "    color=[\"predictions\"],\n",
    "    frameon=False,\n",
    "    title=\"Query predictions\",\n",
    "    ax=ax,\n",
    "    alpha=0.7,\n",
    ")\n",
    "\n",
    "ax = sc.pl.umap(\n",
    "    adata_full,\n",
    "    frameon=False,\n",
    "    show=False,\n",
    ")\n",
    "sc.pl.umap(\n",
    "    adata_full[: adata_query.n_obs],\n",
    "    color=[\"celltype\"],\n",
    "    frameon=False,\n",
    "    title=\"Query observed cell types\",\n",
    "    ax=ax,\n",
    "    alpha=0.7,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanvi_local",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
